server/src/main/java/org/elasticsearch/index/mapper/MappingLookup.java

```diff
@@ -165,6 +169,26 @@ private MappingLookup(Mapping mapping,
         this.objectMappers = Collections.unmodifiableMap(objects);
     }
 
+    private MappingLookup(MappingLookup mappingLookup, Collection<RuntimeField> runtimeMappings) {
+        this.mapping = mappingLookup.mapping;
+        this.hasNested = mappingLookup.hasNested;
+        this.indexAnalyzersMap = mappingLookup.indexAnalyzersMap;
+        this.indexTimeLookup = mappingLookup.indexTimeLookup;
+        this.indexTimeScriptMappers = mappingLookup.indexTimeScriptMappers;
+        this.fieldMappers = mappingLookup.fieldMappers;
+        this.objectMappers = mappingLookup.objectMappers;
+        this.fieldTypeLookup = new FieldTypeLookup(mappingLookup.fieldTypeLookup, runtimeMappings);
```

Comment:
This is one way to achieve this. I am happy with it code-wise. I do wonder if it's too heavy to make a copy of MappingLookup and FieldTypeLookup on each shard. I could not find a way to kind of save work between shards, even if we know that the runtime mappings will be the same on each shard. We parse them on the shards, hence we apply there too.

If we are worried about this, we could take a different approach, perhaps less elegant: avoid making copies of these two objects and rather expose new methods on them that allow to overlay runtime fields on the fly, a bit like search execution used to previously do. I am not too sure how this would look but I thought that it would not be ideal code-wise. 

I am curious to know what you think!